File: AppAnalytics_SDK_Flutter/ios/Classes/TalkingDataAppAnalyticsPlugin.m
#import "TalkingDataAppAnalyticsPlugin.h"
#import "TalkingData.h"

@implementation TalkingDataAppAnalyticsPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"TalkingData_AppAnalytics"
            binaryMessenger:[registrar messenger]];
    TalkingDataAppAnalyticsPlugin* instance = [[TalkingDataAppAnalyticsPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}


+(void)pluginSessionStart:(NSString*)session withChannelId:(NSString*)channelID
{
    [TalkingData sessionStarted:session withChannelId:channelID];
}


-(id)checkArgument:(NSDictionary*)argument forKey:(NSString*)key ofType:(Class)clazz
{
    if (key == nil || argument == nil || clazz == nil  ) {
        return nil;
    }
    id arg = [argument objectForKey:key];
    if (arg == nil) {
        return nil;
    }
    if (![arg isKindOfClass:clazz]) {
        return nil;
    }
    return arg;
}


-(TDProfileType)profileTypeConvert:(NSString*)accTypeStr
{
    if ([accTypeStr isEqualToString:@"ANONYMOUS"]) {
        return TDProfileTypeAnonymous;
    }else if ([accTypeStr isEqualToString:@"REGISTERED"]){
        return TDProfileTypeRegistered;
    }else if ([accTypeStr isEqualToString:@"SINA_WEIBO"]){
        return TDProfileTypeSinaWeibo;
    }else if ([accTypeStr isEqualToString:@"QQ"]){
        return TDProfileTypeQQ;
    }else if ([accTypeStr isEqualToString:@"QQ_WEIBO"]){
        return TDProfileTypeTencentWeibo;
    }else if ([accTypeStr isEqualToString:@"ND91"]){
        return TDProfileTypeND91;
    }else if ([accTypeStr isEqualToString:@"WEIXIN"]){
        return TDProfileTypeWeiXin;
    }else if ([accTypeStr isEqualToString:@"TYPE1"]){
        return TDProfileTypeType1;
    }else if ([accTypeStr isEqualToString:@"TYPE2"]){
        return TDProfileTypeType2;
    }else if ([accTypeStr isEqualToString:@"TYPE3"]){
        return TDProfileTypeType3;
    }else if ([accTypeStr isEqualToString:@"TYPE4"]){
        return TDProfileTypeType4;
    }else if ([accTypeStr isEqualToString:@"TYPE5"]){
        return TDProfileTypeType5;
    }else if ([accTypeStr isEqualToString:@"TYPE6"]){
        return TDProfileTypeType6;
    }else if ([accTypeStr isEqualToString:@"TYPE7"]){
        return TDProfileTypeType7;
    }else if ([accTypeStr isEqualToString:@"TYPE8"]){
        return TDProfileTypeType8;
    }else if ([accTypeStr isEqualToString:@"TYPE9"]){
        return TDProfileTypeType9;
    }else if ([accTypeStr isEqualToString:@"TYPE10"]){
        return TDProfileTypeType10;
    }
    return TDProfileTypeAnonymous;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"onPageStart" isEqualToString:call.method]){
        NSString * pageName = [self checkArgument:call.arguments forKey:@"pageName" ofType:[NSString class]];
        if (pageName) {
            [TalkingData trackPageBegin:pageName];
        }
    } else if ([@"onPageEnd" isEqualToString:call.method]){
        NSString * pageName = [self checkArgument:call.arguments forKey:@"pageName" ofType:[NSString class]];
        if (pageName) {
            [TalkingData trackPageEnd:pageName];
        }
    } else  if ([@"onRegister" isEqualToString:call.method]){
        NSString * profileID = [self checkArgument:call.arguments forKey:@"profileID" ofType:[NSString class]];
        NSString * profileType = [self checkArgument:call.arguments forKey:@"profileType" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        if (profileID && profileType && name) {
            TDProfileType acctype = [self profileTypeConvert:profileType];
            [TalkingData onRegister:profileID type:acctype name:name];
        }
    } else if ([@"onLogin" isEqualToString:call.method]){
        NSString * profileID = [self checkArgument:call.arguments forKey:@"profileID" ofType:[NSString class]];
        NSString * profileType = [self checkArgument:call.arguments forKey:@"profileType" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        if (profileID && profileType && name) {
            TDProfileType acctype = [self profileTypeConvert:profileType];
            [TalkingData onLogin:profileID type:acctype name:name];
        }
    } else if ([@"onEvent" isEqualToString:call.method]){
        NSString * eventID = [self checkArgument:call.arguments forKey:@"eventID" ofType:[NSString class]];
        NSString * eventLabel = [self checkArgument:call.arguments forKey:@"eventLabel" ofType:[NSString class]];
        NSDictionary * params = [self checkArgument:call.arguments forKey:@"params" ofType:[NSDictionary class]];
        if (eventID) {
            [TalkingData trackEvent:eventID label:eventLabel parameters:params];
        }
    }else if([@"onEventWithValue" isEqualToString:call.method]){
        NSString * eventID = [self checkArgument:call.arguments forKey:@"eventID" ofType:[NSString class]];
        NSString * eventLabel = [self checkArgument:call.arguments forKey:@"eventLabel" ofType:[NSString class]];
        NSDictionary * params = [self checkArgument:call.arguments forKey:@"params" ofType:[NSDictionary class]];
        NSNumber * value = [self checkArgument:call.arguments forKey:@"value" ofType:[NSNumber class]];

        if (eventID) {
//#error 记得更新最新的.a和.h，然后打开下边的注释。
            [TalkingData trackEvent:eventID label:eventLabel parameters:params value:value.doubleValue];
        }
        
    }  else if ([@"onViewItem" isEqualToString:call.method]){
        NSString* category = [self checkArgument:call.arguments forKey:@"category" ofType:[NSString class]];
        NSString* itemID = [self checkArgument:call.arguments forKey:@"itemID" ofType:[NSString class]];
        NSString* name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        NSNumber* unitPrice = [self checkArgument:call.arguments forKey:@"unitPrice" ofType:[NSNumber class]];
        [TalkingData onViewItem:itemID category:category name:name unitPrice:unitPrice.intValue];
        
    } else if ([@"onAddItemToShoppingCart" isEqualToString:call.method]){
        NSNumber * amount = [self checkArgument:call.arguments forKey:@"amount" ofType:[NSNumber class]];
        NSString * category = [self checkArgument:call.arguments forKey:@"category" ofType:[NSString class]];
        NSString * itemID = [self checkArgument:call.arguments forKey:@"itemID" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        NSNumber * uniprice = [self checkArgument:call.arguments forKey:@"unitPrice" ofType:[NSNumber class]];
        [TalkingData onAddItemToShoppingCart:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
    } else if ([@"onViewShoppingCart" isEqualToString:call.method]){
        NSArray * shoppingCartDetails = [self checkArgument:call.arguments forKey:@"shoppingCartDetails" ofType:[NSArray class]];
        TalkingDataShoppingCart * sc = [TalkingDataShoppingCart createShoppingCart];
        for (NSDictionary* each in shoppingCartDetails) {
            NSNumber * amount = [self checkArgument:each forKey:@"amount" ofType:[NSNumber class]];
            NSString * category = [self checkArgument:each forKey:@"category" ofType:[NSString class]];
            NSString * itemID = [self checkArgument:each forKey:@"itemID" ofType:[NSString class]];
            NSString * name = [self checkArgument:each forKey:@"name" ofType:[NSString class]];
            NSNumber * uniprice = [self checkArgument:each forKey:@"unitPrice" ofType:[NSNumber class]];
            [sc addItem:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
        }
        [TalkingData onViewShoppingCart:sc];
    } else if ([@"onPlaceOrder" isEqualToString:call.method]){
        NSString * orderId = [self checkArgument:call.arguments forKey:@"orderId" ofType:[NSString class]];
        NSString * amount = [self checkArgument:call.arguments forKey:@"amount" ofType:[NSNumber class]];
        NSArray * currencyType = [self checkArgument:call.arguments forKey:@"currencyType" ofType:[NSString class]];
        [TalkingData onPlaceOrder:orderId amount:amount currencyType:currencyType];
        
    } else if ([@"onOrderPaySucc" isEqualToString:call.method]){
        NSString * orderId = [self checkArgument:call.arguments forKey:@"orderId" ofType:[NSString class]];
        NSString * amount = [self checkArgument:call.arguments forKey:@"amount" ofType:[NSNumber class]];
        NSArray * currencyType = [self checkArgument:call.arguments forKey:@"currencyType" ofType:[NSString class]];
        NSArray * paymentType = [self checkArgument:call.arguments forKey:@"paymentType" ofType:[NSString class]];
        [TalkingData onOrderPaySucc:orderId amount:amount currencyType:currencyType paymentType:paymentType];

    } else if ([@"onCancelOrder" isEqualToString:call.method]){
        NSString * orderId = [self checkArgument:call.arguments forKey:@"orderId" ofType:[NSString class]];
        NSString * amount = [self checkArgument:call.arguments forKey:@"amount" ofType:[NSNumber class]];
        NSArray * currencyType = [self checkArgument:call.arguments forKey:@"currencyType" ofType:[NSString class]];
        [TalkingData onCancelOrder:orderId amount:amount currencyType:currencyType];

    } else if ([@"setGlobalKV" isEqualToString:call.method]) {
        NSString * key = [self checkArgument:call.arguments forKey:@"key" ofType:[NSString class]];
        id value = call.arguments[@"value"];
        if (value) {
            [TalkingData setGlobalKV:key value:value];
        }
    }else if ([@"removeGlobalKV" isEqualToString:call.method]){
        NSString * key = [self checkArgument:call.arguments forKey:@"key" ofType:[NSString class]];
        [TalkingData removeGlobalKV:key];
    }else if ([@"getDeviceID" isEqualToString:call.method]){
        NSString * deviceid = [TalkingData getDeviceID];
        result(deviceid);
    }else {
        result(FlutterMethodNotImplemented);
    }
}


@end

File: AppAnalytics_SDK_Flutter/android/src/main/java/com/talkingdata/plugin/appanalytics/TalkingDataAppAnalyticsPlugin.java
package com.talkingdata.plugin.appanalytics;

import android.content.Context;

import com.tendcloud.tenddata.ShoppingCart;
import com.tendcloud.tenddata.TCAgent;
import com.tendcloud.tenddata.TDProfile;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** TalkingDataAppAnalyticsPlugin */
public class TalkingDataAppAnalyticsPlugin implements MethodCallHandler {
  private Context context;

  private TalkingDataAppAnalyticsPlugin(Context context){
    this.context = context;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "TalkingData_AppAnalytics");
    channel.setMethodCallHandler(new TalkingDataAppAnalyticsPlugin(registrar.context().getApplicationContext()));
  }

  public static void init(Context ctx, String appID, String channelID) {
    TCAgent.init(ctx, appID, channelID);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method){
      case "init":
        TCAgent.init(context,(String) call.argument("appID"),(String) call.argument("channelID"));
        break;
      case "getDeviceID":
        result.success(TCAgent.getDeviceId(context));
        break;
      case "getOAID":
        result.success(TCAgent.getOAID(context));
        break;
      case "onPageStart":
        TCAgent.onPageStart(context, (String) call.argument("pageName"));
        break;
      case "onPageEnd":
        TCAgent.onPageEnd(context, (String) call.argument("pageName"));
        break;
      case "onEvent":
        String eventID = call.argument("eventID");
        String eventLabel = call.argument("eventLabel");
        Map params = null;
        if (call.argument("params") instanceof Map){
          params = call.argument("params");
        }
        TCAgent.onEvent(context, eventID, eventLabel, params);
        break;
      case "onEventWithValue":
        Map paramWith = null;
        if (call.argument("params") instanceof Map){
          paramWith = call.argument("params");
        }
        TCAgent.onEvent(context,
                (String)call.argument("eventID"),
                (String)call.argument("eventLabel"),
                paramWith,
                (double)call.argument("value"));
        break;
      case "setGlobalKV":
        String globalKey = call.argument("key");
        Object globalValue = call.argument("value");
        TCAgent.setGlobalKV(globalKey, globalValue);
        break;
      case "removeGlobalKV":
        String key = call.argument("key");
        TCAgent.removeGlobalKV(key);
        break;
      case "onRegister":
        String profileID = call.argument("profileID");
        String profileType = call.argument("profileType");
        String name = call.argument("name");
        TCAgent.onRegister(profileID, TDProfile.ProfileType.valueOf(profileType), name);
        break;
      case "onLogin":
        profileID = call.argument("profileID");
        profileType = call.argument("profileType");
        name = call.argument("name");
        TCAgent.onLogin(profileID, TDProfile.ProfileType.valueOf(profileType), name);
        break;
      case "onPlaceOrder":
        TCAgent.onPlaceOrder(
                (String) call.argument("orderId"),
                (int) call.argument("amount"),
                (String) call.argument("currencyType")
        );
        break;
      case "onCancelOrder":
        TCAgent.onCancelOrder(
                (String) call.argument("orderId"),
                (int) call.argument("amount"),
                (String) call.argument("currencyType")
        );
        break;
      case "onViewShoppingCart":
        ShoppingCart shoppingCart = ShoppingCart.createShoppingCart();
        List<Map> shoppingCartDetails = call.argument("shoppingCartDetails");
        for (int i = 0; i < shoppingCartDetails.size(); i++){
          Map map = shoppingCartDetails.get(i);
          shoppingCart.addItem(
                  (String)map.get("itemID"),
                  (String)map.get("category"),
                  (String)map.get("name"),
                  (int)map.get("unitPrice"),
                  (int)map.get("amount")
          );
        }

        TCAgent.onViewShoppingCart(shoppingCart);
        break;
      case "onAddItemToShoppingCart":
        TCAgent.onAddItemToShoppingCart(
                (String) call.argument("itemID"),
                (String) call.argument("category"),
                (String) call.argument("name"),
                (int) call.argument("unitPrice"),
                (int) call.argument("amount")
        );
        break;
      case "onOrderPaySucc":
        TCAgent.onOrderPaySucc(
                (String) call.argument("orderId"),
                (int) call.argument("amount"),
                (String) call.argument("currencyType"),
                (String) call.argument("paymentType")
        );
        break;
      case "onViewItem":
        TCAgent.onViewItem(
                (String) call.argument("itemID"),
                (String) call.argument("category"),
                (String) call.argument("name"),
                (int) call.argument("unitPrice")
        );
        break;
      default:
        result.notImplemented();
        break;
    }
  }
}
